source-map.bs

```diff
@@ -674,22 +676,87 @@ The index map follows the form of the standard map.  Like the regular source map
+        1. Otherwise,
+            1. If |offsetLine| is less than |previousOffset|[`"line"`], [=optionally report an error=].
+            1. If |offsetLine| is equal to |previousOffset|[`"line"`] and |offsetColumn| is less than |previousOffset|[`"column"`], [=optionally report an error=].
+
+        Note: This part of the decoding algorithm checks that index source map sections do not overlap. While it is
+              expected that generators should not produce index source maps with overlapping sections, source map
+              consumers may choose instead to only check the simpler condition that the sections are ordered and therefore
+              that offsets appear in increasing order.
+
+    1. If |section|[`"map"`] does not [=map/exist=], then report an error.
+    1. Let |decodedSection| be the result of [=decoding a source map=] given |section|[`"map"`] and |baseURL|.
+    1. If the previous step reported an error, [=optionally report an error=] and continue with the next iteration.
+    1. Set |sourceMap|'s [=decoded source map/sources=] to the result of [=list/extending=] |sourceMap|'s [=decoded source map/sources=] with |decodedSection|'s [=decoded source map/sources=].
+    1. Let |offsetMappings| be a new [=list=].
+    1. Let |firstLine| be null.
+    1. [=For each=] |mapping| of |decodedSection|'s [=decoded source map/mappings=]:
+        1. Let |offsetMapping| be a new [=decoded mapping=].
+        1. Set |offsetMapping|'s [=decoded mapping/generatedLine=] to |mapping|'s [=decoded mapping/generatedLine=] + |offsetLine|.
+        1. If |firstLine| is null, set |firstLine| to |offsetMapping|'s [=decoded mapping/generatedLine=].
+        1. If |offsetMapping|'s [=decoded mapping/generatedLine=] is equal to |firstLine|, set |offsetMapping|'s [=decoded mapping/generatedColumn=] to |mapping|'s [=decoded mapping/generatedColumn=] + |offsetColumn|.
+        1. Otherwise, set |offsetMapping|'s [=decoded mapping/generatedColumn=] to |mapping|'s [=decoded mapping/generatedColumn=].
+        1. Set |offsetMapping|'s [=decoded mapping/originalSource=] to |mapping|'s [=decoded mapping/originalSource=].
+        1. Set |offsetMapping|'s [=decoded mapping/originalLine=] to |mapping|'s [=decoded mapping/originalLine=].
+        1. Set |offsetMapping|'s [=decoded mapping/originalColumn=] to |mapping|'s [=decoded mapping/originalColumn=].
+        1. Set |offsetMapping|'s [=decoded mapping/name=] to |mapping|'s [=decoded mapping/name=].
+        1. [=list/Append=] |offsetMapping| to |offsetMappings|.
+    1. Set |sourceMap|'s [=decoded source map/mappings=] to the result of [=list/extending=] |sourceMap|'s [=decoded source map/mappings=] with |offsetMappings|.
+    1. Set |previousOffset| to |section|[`"offset"`].
+    1. Let |sortedMappings| be the result of [=list/sorting=] |offsetMappings| in ascending order, with an item |a| being less than |b| if |a| is [=generated position less than=] |b|.
```

Comment:
Let's do this before the append? They're N distinct regions, no reason to let the sorting algorithm do position guesses in another sections mappings.

Reply:
The reason it's after the append currently is that this `sortedMappings` is only used to check the section overlap constraint, and the appended mappings don't need to actually be sorted. IIUC you were suggesting sorting the mappings too to avoid the extra computation right?

BTW: I'm actually thinking of proposing to remove the "non-overlapping" constraint or weaken how we define it, because this check seems overly complex. Will bring it up at the next meeting.

Reply:
> The reason it's after the append currently is that this sortedMappings is only used to check the section overlap constraint, and the appended mappings don't need to actually be sorted

Right, which means we only need to sort the appended mappings. We know they don't overlap, everything before must be sorted, so we only need to sort these new mappings before we append.

Right now, there's a bug where you sort a list after it's appended to another list, which means the sort operation is discarded. We either sort the full list after appending, or we sort the smaller list before appending.

> IIUC you were suggesting sorting the mappings too to avoid the extra computation right?

Hopefully mappings is never unsorted, and one of the things I think we should change in the future.

> I'm actually thinking of proposing to remove the "non-overlapping" constraint or weaken how we define it, because this check seems overly complex

The constraint is important for mine, Mozilla's, and Sentry's impls. When doing lookups, we just binary search the offsets, then binary search the section's map. Weakening this means we have to search a union, which can be much slower.

Reply:
> Right, which means we only need to sort the appended mappings. We know they don't overlap, everything before must be sorted, so we only need to sort these new mappings before we append.
>
> Right now, there's a bug where you sort a list after it's appended to another list, which means the sort operation is discarded. We either sort the full list after appending, or we sort the smaller list before appending.

I think I'm not quite following. Right now the draft algorithm sorts only the `sortedMappings` (which is a new copy because list sort is functional, not in-place) and the original unsorted `offsetMappings` is appended. So the appended mappings should be in whatever order they came from decoding, which could be unordered.

> The constraint is important for mine, Mozilla's, and Sentry's impls. When doing lookups, we just binary search the offsets, then binary search the section's map. Weakening this means we have to search a union, which can be much slower.

This makes sense, and I'm convinced the ordering constraint should be kept in. I've simplified/clarified it in the latest push to use the previous `{first_0, last_0}` mappings and the current `{first_1, last_1}` mapping as ranges and do the standard `last_0 < first_1` check.

Do you also rely on the mappings being in the range of {currentOffset, nextOffset}? (that is, the next offset must be larger than the last mapping of the current range)

Reply:
The spec is out of order:
```
    1. [=list/Extend=] |sourceMap|'s [=decoded source map/mappings=] with |offsetMappings|.
    1. Set |previousOffset| to |section|[`"offset"`].
    1. Let |sortedMappings| be the result of [=list/sorting=] |offsetMappings| in ascending order, with an item |a| being less than |b| if |a| is [=generated position less than=] |b|.
```

It concats mappings with `offsetMappings`, then sorts `offsetMappings`, and then we throw away `sortedMappings`.

We can either sort then append, or we can append then sort. In the current ordering, I think you intend to append then sort, but you're sorting the wrong list. If we switch the ordering, then it solves both the bug and my suggestion, we'll only sort the new mappings.

Reply:
It's actually intentional that the sorted list is thrown away, as it's only used to determine the first and last items of each sub-map's mappings range. These first and last items are, in turn, only used to check the overlap requirements.

It's not intended to sort the actual mappings because they're not sorted in the ordinary, non-index map algorithm either and it should probably be consistent between the decoding algorithms.